### PR TITLE
Feature: search box for branch name

### PR DESCRIPTION
**Pull request by Zorin95670**

### CODE REVIEW

File: src/models/git/Branch.js [hunk 1]
```diff
@@ -44,5 +51,21 @@ class Branch {
...
+   */
+  compare(anotherBranch) {
+    if (this.current) {
+      return -1;
+    }
+    if (anotherBranch.current) {
+      return 1;
+    }
+    return this.name.localeCompare(anotherBranch.name);
+  }
```
Comment (justinehell) [inline src/models/git/Branch.js:69]:
```suggestion
   * Compare this branch to the provided branch.
   * @param {Branch} providedBranch - provided branch to compare.
   * @return {Number} A negative number if this branch occurs before provided branch; positive if
   * this branch occurs after provided branch; 0 if they are equivalent.
   */
  compare(providedBranch) {
    if (this.current) {
      return -1;
    }
    if (providedBranch.current) {
      return 1;
    }
    return this.name.localeCompare(providedBranch.name);
  }
```

File: tests/unit/models/git/Branch.spec.js [hunk 2]
```diff
@@ -55,4 +59,24 @@ describe('Test class: Branch', () => {
       });
     });
   });
+
+  describe('Test functions', () => {
+    describe('Test function: compare', () => {
+      it('Current branch should be before other branches', () => {
```
Comment (justinehell) [inline tests/unit/models/git/Branch.spec.js:65]:
```suggestion
      it('Should display current branch before other branches', () => {
```

File: tests/unit/models/git/Branch.spec.js [hunk 3]
```diff
@@ -55,4 +59,24 @@ describe('Test class: Branch', () => {
...
+        const currentBranch = new Branch({ current: true });
+        const otherBranch = new Branch({ current: false });
+        expect(currentBranch.compare(otherBranch)).toEqual(-1);
+        expect(otherBranch.compare(currentBranch)).toEqual(1);
+      });
+
+      it('Should compare with alphabetical order', () => {
```
Comment (justinehell) [inline tests/unit/models/git/Branch.spec.js:72]:
```suggestion
      it('Should compare branches in alphabetical order', () => {
```

File: src/components/menu/GitBranchMenu.vue [hunk 4]
```diff
@@ -1,43 +1,68 @@
...
-      <git-branch-expand-list-menu
-        :open="showRemote"
-        :number="branches.remote.length - maxItem"
-        v-if="branches.remote.length > maxItem"
-        @click="openCloseExpandMenu(false)"
-      />
+
+      <template v-if="filteredBranches.local.length === 0 && filteredBranches.remote.length === 0">
+        <q-item>
+          <q-item-section>Nothing to show.</q-item-section>
```
Comment (justinehell) [inline src/components/menu/GitBranchMenu.vue:64]: Maybe use i18n ?

File: src/models/git/Branch.js [hunk 5]
```diff
@@ -8,12 +8,14 @@ class Branch {
    * @param {Boolean} [props.onLocal=false] - Indicate if branch is on local.
    * @param {Boolean} [props.onRemote=false] - Indicate if branch is on remote.
    * @param {String} [props.remote='origin'] - Remote value.
+   * @param {Boolean} [props.current=false] - Indicates if this branch is the current.
```
Comment (k-y0u) [inline src/models/git/Branch.js:11]: Wrong commit name !

File: src/models/git/Branch.js [hunk 6]
```diff
@@ -35,6 +37,11 @@ class Branch {
      * @type {String}
      */
     this.remote = props.remote || 'origin';
+    /**
+     * Indicates if this branch is the current.
```
Comment (k-y0u) [inline src/models/git/Branch.js:41]: `Indicates` => `Indicate`

File: src/components/menu/GitBranchMenu.vue [hunk 7]
```diff
@@ -1,43 +1,68 @@
...
-      <template v-for="(branch, index) in branches.local">
-        <git-branch-item-menu
-          :key="`local_${branch.name}`"
-          :name="branch.name"
-          :full-name="branch.fullName"
-          :current="branch.name === currentBranchName"
-          v-if="showLocal || index < maxItem"
-          :data-cy="`git-menu-branch-local-${branch.name}`"
+      <q-input
+          class="search-bar"
```
Comment (k-y0u) [inline src/components/menu/GitBranchMenu.vue:10]: indent 4 to 2

File: src/components/menu/GitBranchMenu.vue [hunk 8]
```diff
@@ -1,43 +1,68 @@
...
         />
       </template>
-      <git-branch-expand-list-menu
-        :open="showRemote"
-        :number="branches.remote.length - maxItem"
-        v-if="branches.remote.length > maxItem"
-        @click="openCloseExpandMenu(false)"
-      />
+
+      <template v-if="filteredBranches.local.length === 0 && filteredBranches.remote.length === 0">
```
Comment (k-y0u) [inline src/components/menu/GitBranchMenu.vue:62]: could use a computed

File: src/components/menu/GitBranchMenu.vue [hunk 9]
```diff
@@ -60,14 +86,55 @@ defineProps({
...
 
+/**
+ * Focus on the search input.
+ */
+function focusOnSearch() {
+  searchInput.value.focus();
+}
+
+/**
+ * Indicates if branch name contains search text.
```
Comment (k-y0u) [inline src/components/menu/GitBranchMenu.vue:111]: `Indicates` => `Indicate`
`search` => `searched`

File: src/components/menu/GitBranchMenu.vue [hunk 10]
```diff
@@ -60,14 +86,55 @@ defineProps({
...
+ * @param {String} branch - Branch name.
+ * @return {Boolean} Return true if branch contains search text otherwise return false.
```
Comment (k-y0u) [inline src/components/menu/GitBranchMenu.vue:113]: search => searched

File: src/components/menu/GitBranchMenu.vue [hunk 11]
```diff
@@ -60,14 +86,55 @@ defineProps({
...
+ */
+function isSearched(branch) {
+  if (!searchBranch.value || searchBranch.value.trim().length === 0) {
```
Comment (k-y0u) [inline src/components/menu/GitBranchMenu.vue:116]: `searchBranch` => `searchedBranch`

File: src/components/menu/GitBranchMenu.vue [hunk 12]
```diff
@@ -60,14 +86,55 @@ defineProps({
...
+    return true;
+  }
+
+  return searchBranch.value.toLowerCase().split(' ')
+    .filter((searchText) => searchText !== '')
```
Comment (k-y0u) [inline src/components/menu/GitBranchMenu.vue:121]: `searchText` => `searchedText`

File: src/components/menu/GitBranchMenu.vue [hunk 13]
```diff
@@ -60,14 +86,55 @@ defineProps({
...
+    .some((searchText) => branch.toLowerCase().includes(searchText));
+}
+
+/**
+ * Filter all branches with search text.
```
Comment (k-y0u) [inline src/components/menu/GitBranchMenu.vue:126]: `search text` => `searched text`

File: tests/unit/components/menu/GitBranchMenu.spec.js [hunk 14]
```diff
@@ -62,6 +62,102 @@ describe('Test component: GitBranchMenu', () => {
...
+        name: 'remote',
+        onLocal: false,
+        onRemote: true,
+      });
+      const both = new Branch({
+        name: 'both',
+        onLocal: true,
+        onRemote: true,
+      });
+      it('Should invert local value', () => {
```
Comment (k-y0u) [inline tests/unit/components/menu/GitBranchMenu.spec.js:107]: jump line

File: src/components/menu/GitBranchMenu.vue [hunk 15]
```diff
@@ -1,43 +1,69 @@
...
-        <git-branch-item-menu
-          :key="`local_${branch.name}`"
-          :name="branch.name"
-          :full-name="branch.fullName"
-          :current="branch.name === currentBranchName"
-          v-if="showLocal || index < maxItem"
-          :data-cy="`git-menu-branch-local-${branch.name}`"
+      <q-input
+          ref="searchInput"
+          class="search-bar"
```
Comment (justinehell) [inline src/components/menu/GitBranchMenu.vue:10]: unused class

File: cypress/e2e/git/branch_menu.feature [hunk 16]
```diff
@@ -17,3 +17,24 @@ Feature: Test git branch menu
     When I click on "[data-cy=\"git-current-branch\"]"
     Then I expect "[data-cy=\"git-menu-branch-local-master\"]" exists
     And I expect "[data-cy=\"git-menu-branch-local-master\"] [data-cy=\"git-menu-current-branch\"]" exists
+
+  Scenario: Expect to don't have any branch on "no_master" filter
```
Comment (k-y0u) [inline cypress/e2e/git/branch_menu.feature:21]: `to don't` => `Expect to have no branch with ...`

File: cypress/e2e/git/branch_menu.feature [hunk 17]
```diff
@@ -17,3 +17,24 @@ Feature: Test git branch menu
...
+    When I click on "[data-cy=\"git-current-branch\"]"
+    Then I expect "[data-cy=\"git-menu-branch-local-master\"]" exists
+
+    When I set on "[data-cy=\"git-branch-menu-search-bar\"]" text "not_master"
+    Then I expect "[data-cy=\"git-menu-branch-local-master\"]" not exists
+
+  Scenario: Expect to have master branch on "master" filter
```
Comment (k-y0u) [inline cypress/e2e/git/branch_menu.feature:28]: `with ... filter`

File: changelog.md [hunk 18]
```diff
@@ -26,6 +26,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add plugin manager
 - Display current git branch of project in text view
 - Add menu to list git repository branches
+- Add search field for git branches menu
```
Comment (k-y0u) [inline changelog.md:29]: to put at the end of it !

**Review by k-y0u — CHANGES_REQUESTED** [review]

**Review by justinehell — CHANGES_REQUESTED** [review]

File: src/models/git/Branch.js (hunk 6 above)
Comment (Zorin95670) [reply src/models/git/Branch.js:41]: done

File: src/components/menu/GitBranchMenu.vue (hunk 7 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:10]: done

File: src/components/menu/GitBranchMenu.vue (hunk 8 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:62]: done

File: src/components/menu/GitBranchMenu.vue (hunk 9 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:111]: done

File: src/components/menu/GitBranchMenu.vue (hunk 10 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:113]: done

File: src/components/menu/GitBranchMenu.vue (hunk 11 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:116]: done

File: src/components/menu/GitBranchMenu.vue (hunk 12 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:121]: done

File: src/components/menu/GitBranchMenu.vue (hunk 13 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:126]: done

File: tests/unit/components/menu/GitBranchMenu.spec.js (hunk 14 above)
Comment (Zorin95670) [reply tests/unit/components/menu/GitBranchMenu.spec.js:107]: done

File: cypress/e2e/git/branch_menu.feature (hunk 16 above)
Comment (Zorin95670) [reply cypress/e2e/git/branch_menu.feature:21]: done

File: cypress/e2e/git/branch_menu.feature (hunk 17 above)
Comment (Zorin95670) [reply cypress/e2e/git/branch_menu.feature:28]: done

File: cypress/e2e/git/branch_menu.feature [hunk 19]
```diff
@@ -17,3 +17,24 @@ Feature: Test git branch menu
...
+    When I click on "[data-cy=\"git-current-branch\"]"
+    Then I expect "[data-cy=\"git-menu-branch-local-master\"]" exists
+
+    When I set on "[data-cy=\"git-branch-menu-search-bar\"]" text "master"
+    Then I expect "[data-cy=\"git-menu-branch-local-master\"]" exists
+
+  Scenario: Expect to have master branch on "no_master master" filter
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply cypress/e2e/git/branch_menu.feature:35]: done

File: changelog.md (hunk 18 above)
Comment (Zorin95670) [reply changelog.md:29]: like the previous pr, don't care, I will rework on the changelog before release version

File: src/models/git/Branch.js (hunk 1 above)
Comment (Zorin95670) [reply src/models/git/Branch.js:69]: I propose an alternative :eyes:

File: tests/unit/models/git/Branch.spec.js (hunk 2 above)
Comment (Zorin95670) [reply tests/unit/models/git/Branch.spec.js:65]: done

File: tests/unit/models/git/Branch.spec.js (hunk 3 above)
Comment (Zorin95670) [reply tests/unit/models/git/Branch.spec.js:72]: done

File: src/components/menu/GitBranchMenu.vue (hunk 4 above)
Comment (Zorin95670) [reply src/components/menu/GitBranchMenu.vue:64]: OH yes sorry

File: src/components/menu/GitBranchMenu.vue [hunk 20]
```diff
@@ -1,57 +1,87 @@
...
         />
       </template>
-      <git-branch-expand-list-menu
-        :open="showRemote"
-        :number="branches.remote.length - maxItem"
-        v-if="branches.remote.length > maxItem"
-        @click="openCloseExpandMenu(false)"
-      />
+
+      <template v-if="hasBranches">
```
Comment (justinehell) [inline src/components/menu/GitBranchMenu.vue:61]: Confusing computed name
`hasBranches` => `hasNoBranches` / `noBranches` ?

File: src/components/menu/GitBranchMenu.vue [hunk 21]
```diff
@@ -60,14 +90,57 @@ defineProps({
...
+});
 const branches = ref({
   local: [],
   remote: [],
 });
+const searchedBranch = ref('');
 const showLocal = ref(false);
 const showRemote = ref(false);
+const searchInput = ref(null);
+const hasBranches = computed(() => filteredBranches.value.local.length === 0
```
Comment (justinehell) [inline src/components/menu/GitBranchMenu.vue:105]: Confusing computed name
`hasBranches` => `hasNoBranches` / `noBranches` ?

**Review by justinehell — CHANGES_REQUESTED** [review]

File: src/components/menu/GitBranchMenu.vue (hunk 7 above)
Comment (k-y0u) [reply src/components/menu/GitBranchMenu.vue]: problem still present l.34

File: cypress/e2e/git/branch_menu.feature [hunk 22]
```diff
@@ -17,3 +17,51 @@ Feature: Test git branch menu
...
+    Then I expect "[data-cy=\"git-menu-branch-local-master\"]" exists
+
+  Scenario: Expect to have master branch with "no_master master" filter
+    When I click on "[data-cy=\"git-current-branch\"]"
+    Then I expect "[data-cy=\"git-menu-branch-local-master\"]" exists
+
+    When I set on "[data-cy=\"git-branch-menu-search-bar\"]" text "no_master master"
+    Then I expect "[data-cy=\"git-menu-branch-local-master\"]" exists
+
+    Scenario: Save git configuration twice should not duplicate branches in branch menu
```
Comment (k-y0u) [inline cypress/e2e/git/branch_menu.feature:42]: it feels like a bad indentation

**Review by k-y0u — APPROVED** [review]

**Review by justinehell — APPROVED** [review]

**Pull request merged**